### PR TITLE
CHOP-42. Cleanup and add tests for doAndRelease, doAndClose.

### PR DESCRIPTION
**Pull request by robotoer**
@lockoff.

### CODE REVIEW

File: src/main/scala/org/kiji/chopsticks/DSL.scala [hunk 1]
```diff
@@ -27,6 +29,14 @@ import org.kiji.schema.util.ReferenceCountable
 @ApiAudience.Public
 @ApiStability.Unstable
 object DSL {
+  /**
+   * Exception that contains multiple exceptions.
+   *
+   * @param msg Message to include with the exception.
+   * @param errors Multiple exceptions causing this exception.
+   */
+  final case class CompoundException(msg: String, errors: Seq[Throwable]) extends Exception
```
Comment (lockoff) [inline src/main/scala/org/kiji/chopsticks/DSL.scala]: This should probably be a sequence of Exception and not Throwable. The only other kind of Throwable is Error, and an Error should never be caught or dealt with by developers.

File: src/main/scala/org/kiji/chopsticks/NewsgroupLoader.scala [hunk 2]
```diff
@@ -98,29 +100,22 @@ object NewsgroupLoader {
...
         newsgroup <- root.listFiles()
-        posting <- newsgroup.listFiles()
-        word <- getWords(posting)
-      } writer.get.put(table.get.getEntityId(word), "info", "word", word)
-    } finally {
-      // Cleanup connections.
-      kiji.foreach { ResourceUtils.releaseOrLog(_) }
-      table.foreach { ResourceUtils.closeOrLog(_) }
-      writer.foreach { ResourceUtils.closeOrLog(_) }
+      } yield future {
```
Comment (lockoff) [inline src/main/scala/org/kiji/chopsticks/NewsgroupLoader.scala:106]: I'm a bit confused why futures have come into play.  Why was this necessary?

File: src/main/scala/org/kiji/chopsticks/DSL.scala [hunk 3]
```diff
@@ -27,6 +29,14 @@ import org.kiji.schema.util.ReferenceCountable
 @ApiAudience.Public
 @ApiStability.Unstable
 object DSL {
+  /**
```
Comment (lockoff) [inline src/main/scala/org/kiji/chopsticks/DSL.scala]: At this point I think this functionality should be moved into another file, since they're not limited to the domain-specific language we're making with Scalding. Eventually this sort of utility code could even go in another project of tools for Scala developers (but another file in this project is cool for now).  Can we make a utility class or object to hold this functionality?

File: src/main/scala/org/kiji/chopsticks/NewsgroupLoader.scala (hunk 2 above)
Comment (robotoer) [inline src/main/scala/org/kiji/chopsticks/NewsgroupLoader.scala:106]: This makes NewsgroupLoader multi-threaded and much faster as a result.

File: src/main/scala/org/kiji/chopsticks/DSL.scala (hunk 3 above)
Comment (robotoer) [inline src/main/scala/org/kiji/chopsticks/DSL.scala]: I had that thought as well but was lazy. I'll make a new object for these. Any preference on the name/location?

File: src/main/scala/org/kiji/chopsticks/DSL.scala (hunk 1 above)
Comment (robotoer) [inline src/main/scala/org/kiji/chopsticks/DSL.scala]: Changed!

File: src/main/scala/org/kiji/chopsticks/NewsgroupLoader.scala (hunk 2 above)
Comment (lockoff) [inline src/main/scala/org/kiji/chopsticks/NewsgroupLoader.scala:106]: Ooooh fun.

File: src/main/scala/org/kiji/chopsticks/Resources.scala [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+@ApiAudience.Public
+@ApiStability.Unstable
+object Resources {
+  /**
+   * Exception that contains multiple exceptions.
+   *
+   * @param msg Message to include with the exception.
+   * @param errors Multiple exceptions causing this exception.
+   */
+  final case class CompoundException(msg: String, errors: Seq[Exception]) extends Exception
```
Comment (lockoff) [inline src/main/scala/org/kiji/chopsticks/Resources.scala:37]: perhaps this kind of exception should be private.  Do we want end users making instances of this exception themselves?

File: src/main/scala/org/kiji/chopsticks/Resources.scala [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+  /**
+   * Performs an operation with a resource that requires post processing. This method will throw a
+   * CompoundException when exceptions get thrown during the operation and while resources are
```
Comment (lockoff) [inline src/main/scala/org/kiji/chopsticks/Resources.scala:41]: docs link on CompoundException?

**Comment by lockoff** [comment]
Great stuff.  @robotoer let me know if the changes I've commented on are cool, and I'll make them and then push.

File: src/main/scala/org/kiji/chopsticks/Resources.scala (hunk 4 above)
Comment (robotoer) [inline src/main/scala/org/kiji/chopsticks/Resources.scala:37]: Wouldn't clients want to be able to catch this type of exception to get its children exceptions?

File: src/main/scala/org/kiji/chopsticks/Resources.scala (hunk 5 above)
Comment (robotoer) [inline src/main/scala/org/kiji/chopsticks/Resources.scala:41]: Oops.

Comment (robotoer) [inline src/main/scala/org/kiji/chopsticks/Resources.scala:41]: Do scaladocs support the @throws tag?

File: src/main/scala/org/kiji/chopsticks/Resources.scala (hunk 4 above)
Comment (lockoff) [inline src/main/scala/org/kiji/chopsticks/Resources.scala:37]: oh blergh duh.

File: src/main/scala/org/kiji/chopsticks/Resources.scala (hunk 5 above)
Comment (lockoff) [inline src/main/scala/org/kiji/chopsticks/Resources.scala:41]: Yes, check out the [Scaladocs Wiki](https://wiki.scala-lang.org/display/SW/Tags+and+Annotations).
